doc/design.md

```diff
@@ -0,0 +1,949 @@
+
+The library is intended as a low-level building block and the mixed-multi-dimensional aspected lead to many corner cases.
+A comprehensive and maintainable suite of unit tests is thus essential.
+
+
+### <a name="design-details-benchmarks"></a>Benchmarks
+
+Given the intention of being a low-level building block for everything else, ensuring good performance and preventing performance regression is thus essential.
+A suite of benchmarks is to be maintained and monitored for changes.
+
+
+### <a name="design-details-sanitizers"></a>Sanitizers
+
+Static analysis and sanitization runs should be performed and verified automatically.
+Most importantly, AddressSanitizer should be used to strictly monitor for memory corruption bugs.
+
+
+### <a name="design-details-code-review"></a>Code review
+
+The same review/approval system for pull request as in Mantid should be adopted.
+However, given the scope of the project, we should require *two full reviews*, i.e., the gatekeeper should also do a thorough review, in addition to the first reviewer.
+
+
+### <a name="design-details-documentation"></a>Documentation
+
+Documentation is required for several cases:
+- Library (developer) documentation.
+- C++ API (user) documentation.
+- Python API (user) documentation.
+- Python usage examples beyond simple API documentation, e.g., how to represent complex data structures such as a `DataObjects::Workspace2D` as a `Dataset`.
+
+
+### <a name="design-details-scope-and-extension"></a>Scope and extension
+
+The design described here is seen as a nearly complete description of the scope of the dataset library.
+Extending the functionally should be avoided and should happen only after careful consideration.
+In particular, anything that is *more* specific than already supported should not be made part of this library, but instead be placed in another library that depends on this library.
+- Actually `convert` to convert units is pretty specific as it needs to connect instrument information to data.
+  Is having it as part of this lowest-level library the right choice?
+
+There are some cases which are likely to lead to additions to the dataset library:
+- Initially we will not have covered all required cases with tags for `Dim` and variable tags for `Coord` and `Data`.
+  In particular, we need a new `Coord` for every unit that we support.
+- As a consequence the unit handling will also need to be extended.
+
+
+### <a name="design-details-api-stability"></a>API stability
+
+The C++ version as well as the Python exports are meant as a low level building block for other parts of the project, in addition to direct use in scripts by users.
+Therefore we need to ensure that the API is reasonably stable.
+At the same time, we will struggle with things that do not work out early on as imagined right now and will need to change the API.
+- The API is considered unstable and can change without notice until the library version has reached v1.0.
```

Comment:
So this should tie in to Milestones somehow. i.e. need to be at >1 before migrating.

Reply:
Migrating what?